### ISSUE TITLE
Add a file tree abstraction for mod installers (and other parts of the app)

### ISSUE THREAD

**Issue by halgari** [Nexus-Mods/NexusMods.App#553]
# Improvement

## Currently 
Files are passed into installers as a `Dictionary<RelativePath, Hash>` while this is a very simple interface each installer ends up writing their own tree traversal code.

## We want
To consolidate common operations on these entries into a simple data model. This file tree class should:

* Be somewhat path agnostic, should work for relative and absolute paths
* Support:
  * Finding the parent of an entry
  * Finding sibblings of an entry
  * Finding all decendents of an entry
  * Getting information of an entry (hash, size, etc)
* Recommended: keep the tree as a generic, so that different "value entry" types can be provided, so we can quickly convert a `Dictionary<RelativePath, Hash>` or a `Dictionary<GamePath, FileEntry>` to a `FileTree<TPath, TValue>`
* May require implementing a more detailed `IPath` interface

**Comment by Al12rs** [comment]
Mod Organizer 2 has an IFileTree abstraction that it uses extensively in installers and throughout the app. 
https://github.com/ModOrganizer2/modorganizer-uibase/blob/b3009263a7614fc96c14ef1329d095f05c84d167/src/ifiletree.h#L70

We have different implementations of the abstraction, some referencing real filesystem file structures, other indicating "virtual" trees.

https://github.com/ModOrganizer2/modorganizer/blob/c8e247f900836b58abdb93f130629c91ed4be0c2/src/archivefiletree.cpp#L77

https://github.com/ModOrganizer2/modorganizer/blob/c8e247f900836b58abdb93f130629c91ed4be0c2/src/virtualfiletree.cpp#L17

**Comment by Al12rs** [comment]
I don't think we actually have any utility to differentiate an `AbsolutePath` / `RelativePath` to a File rather than to a Directory.
We usually only deal with files, but a `RelativePath` can just as well indicate a directory.

If we know that the input to generate the tree is a list of purely files then, it is possible to correctly subdivide them between files and directories. 

Otherwise we could have a single `Children` property that returns all the child entries, then have a HasChildren() to easily check whether it has further entries. If it has, you know its a directory, if it doesn't you can only guess.
